File: junit4-custom-retry-runner/src/test/java/retries/RetryableTests.java
package retries;
import core.Junit4Rerunner;
import org.junit.*;
import org.junit.runner.RunWith;

@RunWith(Junit4Rerunner.class)
public class RetryableTests {

    private int tryCount = 0;

    @BeforeClass
    public static void beforeAll() {
        System.out.println("BeforeAll");
    }

    @AfterClass
    public static void afterAll() {
        System.out.println("AfterAll");
    }

    @Before
    public void beforeEach() {
        System.out.println("BeforeEach");
    }

    @After
    public void afterEach() {
        System.out.println("AfterEach");
    }

    @Test
    public void failedTest() {
        System.out.println("testFailed");
        throw new RuntimeException("FAIL");
    }

    @Test
    public void successTest() {
        System.out.println("testSuccess");
    }

    @Test
    public void successAfterRetryTest() {
        if (tryCount < 2) {
            tryCount++;
            System.out.println("testFailed");
            throw new RuntimeException("testFailed");
        }
        System.out.println("testSuccess");
    }
}

File: kotlin-retry-wrapper/src/test/kotlin/RetryableTests.kt
import core.retryableTest
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test


@DisplayName("Дожатие тестов с помощью обертки над тестом")
class RetryableTest {

    var tryCount = 0

    @Test
    fun failedTest() = retryableTest {
        println("testFailed")
        throw Exception("FAIL")
    }

    @Test
    fun successTest() = retryableTest {
        println("testSuccess")
    }

    @Test
    fun successAfterRetryTest() = retryableTest {
        if (tryCount < 2) {
            tryCount++
            println("testFailed")
            throw Exception("testFailed")
        }
        println("testSuccess")
    }
}